### PR TITLE
compiler: more progress on incremental

### PR DESCRIPTION
**Pull request by mlugg**
We're deep in the weeds now. As of this PR trivial updates are blocked on linker progress (or CBE/LLVM backend progress), but I can still make progress in the meantime.

Before I make more steps towards incremental, I need to have a stern talking-to with `Zcu.Decl` (see [my proposed refactor](https://gist.github.com/mlugg/d0dcb0025dbfd7145b66d0ada6060024)) - but I think we're very close to something usable on the frontend.

### CODE REVIEW

File: src/Module.zig [hunk 1]
```diff
@@ -4103,6 +4103,21 @@ pub fn scanNamespace(
     for (decls) |decl_inst| {
         try scanDecl(&scan_decl_iter, decl_inst);
     }
+
+    if (seen_decls.count() != namespace.decls.count()) {
+        // Do a pass over the namespace contents and remove any decls from the last update
+        // which were removed in this one.
+        var i: usize = 0;
+        while (i < namespace.decls.count()) {
```
Comment (jinroh) [inline src/Module.zig:4289]: missing `i += 1` ?

Comment (nektro) [reply src/Module.zig:4289]: 4294 should make it unnecessary but 4295 should immediately crash

Comment (mlugg) [reply src/Module.zig:4289]: Oops, you're totally right, good spot! Luckily this doesn't affect code built without incremental, and this is going to be changed a lot again soon anyway, but thank you for pointing out the issue.

**Comment by andrewrk** [comment]
In case you didn't think of it already, a nice stepping stone might be an incremental `-fno-emit-bin` build. No backend or linker coordination needed, but you can still make the frontend save its work as far as semantic analysis and state serialization/restoration. In fact this configuration might be the very one that IDEs are interested in - getting analysis results quickly with no codegen.

**Comment by mlugg** [comment]
Yeah, that thought has crossed my mind. I need to step back and do some prereq work for now - i.e. mutilating comptime-mutable memory and `Decl` - but once I get back to this I'll probably look into doing `-fno-emit-bin` incremental first. Once that's usable, we could probably look more seriously at #615.

File: src/Module.zig [hunk 2]
```diff
@@ -4103,6 +4103,21 @@ pub fn scanNamespace(
...
+            const decl_index = namespace.decls.keys()[i];
+            const decl = zcu.declPtr(decl_index);
+            if (!seen_decls.contains(decl.name)) {
+                // We must preserve namespace ordering for @typeInfo.
+                namespace.decls.orderedRemoveAt(i);
```
Comment (andrewrk) [inline src/Module.zig:4294]: you can't put a orderedRemove() inside a loop. That's O(N)

Comment (mlugg) [reply src/Module.zig:4294]: Oops, apologies - I didn't think of that. This code is only hit in the incremental path, and it'll probably be changed again anyway after the `Decl` changes, but I'll make sure this gets fixed!

Comment (andrewrk) [reply src/Module.zig:4294]: Thanks!

**Pull request merged**